crates/btc-notify/src/lib.rs

```diff
@@ -0,0 +1,1233 @@
+                    lifecycle.block = Some(blockhash);
+                    diff.push((matched_tx.clone(), TxStatus::Mined));
+                }
+            }
+        }
+
+        if self.unburied_blocks.len() > self.bury_depth {
+            if let Some(newly_buried) = self.unburied_blocks.pop_back() {
+                // Now that we've handled the Mined transitions. We can take the oldest block we are still tracking and
+                // declare all of its relevant transactions buried, and then finally we can clear the buried
+                // transactions from the current lifecycle map.
+                for buried_tx in newly_buried.txdata {
+                    self.tx_lifecycles.remove(&buried_tx.compute_txid());
+                    if self.tx_filters.iter().any(|f|f(&buried_tx)) {
+                        diff.push((buried_tx, TxStatus::Buried));
+                    }
+                }
+            }
+        }
+
+        diff
+    }
+
+    // process_tx is one of the three primary state transition functions of the BtcZmqSM, updating internal state to
+    // reflect the contents of the transaction.
+    fn process_tx(&mut self, tx: Transaction) -> Vec<(Transaction, TxStatus)> {
+        if !self.tx_filters.iter().any(|f|f(&tx)) {
+            return Vec::new();
+        }
+
+        let txid = tx.compute_txid();
+        let lifecycle = self.tx_lifecycles.get_mut(&txid);
+        match lifecycle {
+            // In this case we have never seen any information on this transaction whatsoever.
+            None => {
+                let lifecycle = TxLifecycle {
+                    raw: tx,
+                    block: None,
+                };
+
+                self.tx_lifecycles.insert(txid, Some(lifecycle));
+
+                // We intentionally DO NOT return the transaction here in the diff because we are unsure of what the
+                // status is. We will either immediately get a followup block or a followup sequence which will cause us
+                // to emit a new state change.
+                Vec::new()
+            }
+            // In this case we have seen a MempoolAcceptance event for this txid, but haven't seen the actual
+            // transaction data yet.
+            Some(None) => {
+                let lifecycle = TxLifecycle {
+                    raw: tx.clone(),
+                    block: None,
+                };
+
+                self.tx_lifecycles.insert(txid, Some(lifecycle));
+
+                // Presence within the map indicates we have already received the sequence message for this but don't
+                // yet have any other information, indicating that this rawtx event can generate the Mempool event.
+                vec![(tx, TxStatus::Mempool)]
+            }
+            // In this case we know everything we need to about this transaction, and this is probably a rawtx event
+            // that accompanies an upcoming new block event.
+            Some(Some(_)) => Vec::new(),
```

Comment:
Same idea with the optional return value.
You wouldn't need to return an allocated vec.